File: TAFGENIUSMobile/lib/backend/firestore/services/user_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../../../Model/user_model.dart';

class UserService {
  final CollectionReference usersCollection =
  FirebaseFirestore.instance.collection('users');

  // -------- CRUD --------
  Future<void> createUser(UserModel user) async {
    await usersCollection.doc(user.id).set(user.toMap());
  }

  Future<UserModel?> getUser(String userId) async {
    final doc = await usersCollection.doc(userId).get();
    if (!doc.exists) return null;
    return UserModel.fromFirestore(doc);
  }

  Future<void> updateUser(String userId, Map<String, dynamic> data) async {
    await usersCollection.doc(userId).update(data);
  }

  Future<void> deleteUser(String userId) async {
    await usersCollection.doc(userId).delete();
  }

  // -------- Progrès --------
  Future<void> updateCourseProgress(
      String userId, String courseId, Progress progress) async {
    await usersCollection.doc(userId).update({
      "progress.$courseId": progress.toMap(),
    });
    await _updateGlobalProgress(userId);
  }

  Future<Progress?> getCourseProgress(String userId, String courseId) async {
    final doc = await usersCollection.doc(userId).get();
    if (!doc.exists) return null;

    final progressMap = (doc.data() as Map<String, dynamic>)["progress"];
    if (progressMap == null || !progressMap.containsKey(courseId)) return null;

    return Progress.fromMap(progressMap[courseId]);
  }

  // -------- Calculer le progrès global --------
  Future<void> _updateGlobalProgress(String userId) async {
    final doc = await usersCollection.doc(userId).get();
    if (!doc.exists) return;

    final progressMap = (doc.data() as Map<String, dynamic>)["progress"];
    if (progressMap == null || progressMap.isEmpty) return;

    double total = 0;
    progressMap.forEach((key, value) {
      total += (value['percent'] ?? 0);
    });
    final globalProgress = total / progressMap.length;

    await usersCollection.doc(userId).update({"globalProgress": globalProgress});
  }

  // -------- Liste complète des utilisateurs --------
  Future<List<UserModel>> getAllUsers() async {
    final snapshot = await usersCollection.orderBy('createdAt').get();
    return snapshot.docs
        .map((doc) => UserModel.fromFirestore(doc))
        .toList();
  }

  // -------- Recherche par rôle --------
  Future<List<UserModel>> getUsersByRole(String role) async {
    final snapshot = await usersCollection.where('role', isEqualTo: role).get();
    return snapshot.docs.map((doc) => UserModel.fromFirestore(doc)).toList();
  }

  // -------- Filtrer par nom/email --------
  Future<List<UserModel>> searchUsers(String query) async {
    final snapshot = await usersCollection.get();
    final lowerQuery = query.toLowerCase();

    return snapshot.docs
        .map((doc) => UserModel.fromFirestore(doc))
        .where((user) =>
    user.name.toLowerCase().contains(lowerQuery) ||
        user.email.toLowerCase().contains(lowerQuery))
        .toList();
  }
}

File: src/backend/firestore/services/userService.js
import {
  collection,
  doc,
  getDoc,
  getDocs,
  setDoc,
  updateDoc,
  deleteDoc,
  query,
  where,
  orderBy,
} from "firebase/firestore";

import { db } from "../../../firebaseConfig.js";
import { UserModel } from "../../../model/UserModel.js";
import { Progress } from "../../../model/UserModel.js";

export const UserService = {
  // -------- CRUD --------

  async createUser(user) {
    await setDoc(doc(db, "users", user.id), user.toMap());
  },

  async getUser(userId) {
    const snapshot = await getDoc(doc(db, "users", userId));
    if (!snapshot.exists()) return null;
    return UserModel.fromFirestore(snapshot);
  },

  async updateUser(userId, data) {
    await updateDoc(doc(db, "users", userId), data);
  },

  async deleteUser(userId) {
    await deleteDoc(doc(db, "users", userId));
  },

  // -------- Progrès --------

  async updateCourseProgress(userId, courseId, progress) {
    // progress est une instance de Progress
    await updateDoc(doc(db, "users", userId), {
      [`progress.${courseId}`]: progress.toMap(),
    });

    await UserService._updateGlobalProgress(userId);
  },

  async getCourseProgress(userId, courseId) {
    const snapshot = await getDoc(doc(db, "users", userId));
    if (!snapshot.exists()) return null;

    const data = snapshot.data();
    const progressMap = data?.progress;

    if (!progressMap || !progressMap[courseId]) return null;

    return Progress.fromMap(progressMap[courseId]);
  },

  // -------- Calculer le progrès global --------

  async _updateGlobalProgress(userId) {
    const snapshot = await getDoc(doc(db, "users", userId));
    if (!snapshot.exists()) return;

    const data = snapshot.data();
    const progressMap = data?.progress;

    if (!progressMap || Object.keys(progressMap).length === 0) return;

    let total = 0;
    const keys = Object.keys(progressMap);

    keys.forEach((id) => {
      total += progressMap[id]?.percent || 0;
    });

    const globalProgress = total / keys.length;

    await updateDoc(doc(db, "users", userId), {
      globalProgress,
    });
  },

  // -------- Liste complète des utilisateurs --------

  async getAllUsers() {
    const q = query(collection(db, "users"), orderBy("createdAt"));
    const snapshot = await getDocs(q);

    return snapshot.docs.map((docSnap) => UserModel.fromFirestore(docSnap));
  },

  // -------- Recherche par rôle --------

  async getUsersByRole(role) {
    const q = query(
      collection(db, "users"),
      where("role", "==", role)
    );

    const snapshot = await getDocs(q);
    return snapshot.docs.map((docSnap) => UserModel.fromFirestore(docSnap));
  },

  // -------- Filtrer par nom/email --------

  async searchUsers(searchQuery) {
    const snapshot = await getDocs(collection(db, "users"));
    const lowerQuery = searchQuery.toLowerCase();

    return snapshot.docs
      .map((docSnap) => UserModel.fromFirestore(docSnap))
      .filter(
        (user) =>
          user.name.toLowerCase().includes(lowerQuery) ||
          user.email.toLowerCase().includes(lowerQuery)
      );
  },
};
